packages/core/src/sync/index.ts

```diff
@@ -280,117 +356,155 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
+          getChainsCheckpoint("end")! < getChainsCheckpoint("finalized")!
+            ? getChainsCheckpoint("end")!
+            : getChainsCheckpoint("finalized")!;
+        const to = getChainsCheckpoint("latest") ?? end;
+
+        /*
+         * Extract events with `syncStore.getEvents()`, paginating to
+         * avoid loading too many events into memory.
+         */
+        while (true) {
+          if (isKilled) return;
+          if (from === to) break;
+          const getEventsMaxBatchSize = args.common.options.syncEventsQuerySize;
+          let consecutiveErrors = 0;
+
+          // convert `estimateSeconds` to checkpoint
+          const estimatedTo = encodeCheckpoint({
+            ...zeroCheckpoint,
+            blockTimestamp: Math.min(
+              decodeCheckpoint(from).blockTimestamp + estimateSeconds,
+              maxCheckpoint.blockTimestamp,
+            ),
+          });
 
+          try {
+            const { events, cursor } = await args.syncStore.getEvents({
+              filters: args.sources.map(({ filter }) => filter),
+              from,
+              to: to < estimatedTo ? to : estimatedTo,
+              limit: getEventsMaxBatchSize,
+            });
+            consecutiveErrors = 0;
+
+            for (const network of args.networks) {
+              updateStatus({ events, checkpoint: cursor, network });
+            }
+
+            estimateSeconds = estimate({
+              from: decodeCheckpoint(from).blockTimestamp,
+              to: decodeCheckpoint(cursor).blockTimestamp,
+              target: getEventsMaxBatchSize,
+              result: events.length,
+              min: 10,
+              max: 86_400,
+              prev: estimateSeconds,
+              maxIncrease: 1.08,
+            });
+
+            yield { events, checkpoint: to };
+            from = cursor;
+          } catch (error) {
+            // Handle errors by reducing the requested range by 10x
+            estimateSeconds = Math.max(10, Math.round(estimateSeconds / 10));
+            if (++consecutiveErrors > 4) throw error;
+          }
+        }
+      }
+
+      /** `true` if all networks have synced all known finalized blocks.  */
+      const allHistoricalSyncExhaustive = Array.from(
+        localSyncData.values(),
+      ).every(({ blockProgress }) => {
+        if (isSyncExhaustive(blockProgress)) return true;
+
+        // Determine if `finalized` block is considered "stale"
+        const staleSeconds = (Date.now() - latestFinalizedFetch) / 1_000;
+        if (staleSeconds <= args.common.options.syncHandoffStaleSeconds) {
+          return true;
+        }
+
+        return false;
+      });
+
+      if (allHistoricalSyncExhaustive) break;
+
+      /**
+       * At least one network has a `finalized` block that is considered "stale".
+       */
+
+      latestFinalizedFetch = Date.now();
+
+      await Promise.all(
+        Array.from(localSyncData.entries()).map(
```

Comment:
Note that you could skip refetching finalized block for any chains that are exhaustive/completed here.